omp/components/bitvector.hpp

```diff
@@ -0,0 +1,124 @@
+    array<IndexType> rank_array{exec, num_blocks};
+    const auto bits = bit_array.get_data();
+    const auto ranks = rank_array.get_data();
+#pragma omp parallel for
+    for (IndexType block_i = 0; block_i < num_blocks; block_i++) {
+        const auto base_i = block_i * block_size;
+        storage_type mask{};
+        const auto local_op = [&](int local_i) {
+            const storage_type bit = device_predicate(base_i + local_i) ? 1 : 0;
+            mask |= bit << local_i;
+        };
+        if (base_i + block_size <= size) {
+            for (int local_i = 0; local_i < block_size; local_i++) {
+                local_op(local_i);
+            }
+        } else {
+            for (int local_i = 0; base_i + local_i < size; local_i++) {
+                local_op(local_i);
+            }
+        }
+        bits[block_i] = mask;
+        ranks[block_i] = gko::detail::popcount(mask);
+    }
+    components::prefix_sum_nonnegative(exec, ranks, num_blocks);
+
+    return gko::bitvector<IndexType>{std::move(bit_array),
+                                     std::move(rank_array), size};
+}
+
+
+template <typename IndexIterator>
+gko::bitvector<typename std::iterator_traits<IndexIterator>::value_type>
+from_sorted_indices(
+    std::shared_ptr<const DefaultExecutor> exec, IndexIterator it,
+    typename std::iterator_traits<IndexIterator>::difference_type count,
+    typename std::iterator_traits<IndexIterator>::value_type size)
+{
+    using index_type = typename std::iterator_traits<IndexIterator>::value_type;
+    using bv = device_bitvector<index_type>;
+    using storage_type = typename bv::storage_type;
+    constexpr auto block_size = bv::block_size;
+    const auto num_blocks = static_cast<size_type>(ceildiv(size, block_size));
+    array<storage_type> bit_array{exec, num_blocks};
+    array<index_type> rank_array{exec, num_blocks};
+    const auto bits = bit_array.get_data();
+    const auto ranks = rank_array.get_data();
+    components::fill_array(exec, bits, num_blocks, storage_type{});
+    const auto num_threads = omp_get_max_threads();
+    const auto work_per_thread = ceildiv(count, num_threads);
+    assert(std::is_sorted(it, it + count));
+#pragma omp parallel num_threads(num_threads)
+    {
+        const auto tid = omp_get_thread_num();
+        const auto begin = std::min<index_type>(tid * work_per_thread, count);
+        const auto end = std::min<index_type>(begin + work_per_thread, count);
+        if (begin < end) {
+            const auto first_block = it[begin] / block_size;
+            const auto last_block = it[end - 1] / block_size;
+            storage_type word{};
+            auto block = first_block;
+            for (auto i : irange{begin, end}) {
+                const auto value = it[i];
+                const auto new_block = value / block_size;
+                const auto local = value % block_size;
+                if (new_block != block) {
+                    assert(new_block > block);
```

Comment:
I think it assumes no duplicated index to make this assert.

Reply:
No, this assumes sortedness. I assert that there are no duplicates in other places though